### ISSUE TITLE
router hit 10k fd limit

### ISSUE THREAD

**Issue by scottrobertson** [flynn/flynn#3470]
Hey

I have tried flynn 3 times over the past few months. Each time has resulted in me going back to another service due to Flynn going down quite often. 

This is hosted on 3x 16GB Digital Ocean servers. It was installed using the cloud installer.

The last time this happened, the dashboard went down along with my apps, and i assumed it was a DNS issue, so i migrated the DNS and it came back up. I assume this was due to something being restarted internally, as it did it again a few days later. This triggered me to move away as it was causing downtime.

This time, I received a pingdom notification a few moments ago that the site was down. I checked and the dashboard was not accessible, along with the apps.They came back up on their own, but i suspect they will go down again soon. 

Logs: http://d.pr/f/FOks

Thanks

**Comment by scottrobertson** [comment]
Update: It's just gone down again

**Comment by titanous** [comment]
The error from the router is:

> dial tcp 100.100.76.21:8080: socket: too many open files

Try running this command:

```text
flynn -a router limit set app max_fd=100000
```

The default fd limit is 10,000 which should be enough, so we should figure out why the router is running out of file descriptors.

What languages, frameworks, and web servers are your apps built with?

**Comment by scottrobertson** [comment]
Hey @titanous 

Tried running that, and I got:

`21:29:52.864227 main.go:172: no clusters configured`

There are 2 Rails apps running, and a tiny python app. 1 of them is pretty high traffic, and the other has quite high Sidekiq throughput. There are also 2 redis servers running linked to each of the Rails apps, mainly used for Sidekiq.

**Comment by titanous** [comment]
Do you have the `flynn` CLI configured on your local machine? If so, run the command there.

**Comment by scottrobertson** [comment]
Ah, locally. That worked

**Comment by titanous** [comment]
What webserver are you using for the Rails app? Can you quantify the amount of traffic it gets? (requests per minute, concurrent users, etc.)

I'm trying to figure out whether the high-water mark for the file descriptors is legitimately above 10k or if there's a leak somewhere.

**Comment by scottrobertson** [comment]
Puma for both of the Rails apps.

Errm, i am not 100% sure on how much traffic they get. It's not going to be anywhere near enough to hit that limit though to be honest. It's not even 10+ concurrent connections.

**Comment by titanous** [comment]
Okay, start by gisting the output of this command on each of the hosts, which will provide a (redacted) list of TCP connections and states:

```text
netstat -tuna | sed -r 's/(([0-9]{1,3}\.){3})[0-9]{1,3}/\1x/g'
```

**Comment by scottrobertson** [comment]
Here are the outputs: https://gist.github.com/scottrobertson/d3ebd9e9d70c589ee57fd953961ea06b

**Comment by titanous** [comment]
Oh, changing the limit restarted the routers which cleaned everything up. Wait a few days and lets take another look.

**Comment by scottrobertson** [comment]
Shall do 👍 thanks for the help

**Comment by titanous** [comment]
Happy to help, please let us know if you have any other questions or problems.

**Comment by titanous** [comment]
@scottrobertson Do you mind checking `netstat` again?

Also, while you're there can you run this command:

```text
sudo ls /proc/$(ps xa | grep [f]lynn-router | cut -f1 -d" ")/fd | wc -l
```

**Comment by scottrobertson** [comment]
Here are the updated details: https://gist.github.com/scottrobertson/98c1de4be4e35349212e19f3c9950850

**Comment by titanous** [comment]
Thanks. What are the values of these settings?

```text
sysctl \
  net.ipv4.tcp_keepalive_time \
  net.ipv4.tcp_keepalive_intvl \
  net.ipv4.tcp_keepalive_probes \
  net.netfilter.nf_conntrack_tcp_timeout_established
```

**Comment by scottrobertson** [comment]
They are all:

```
net.ipv4.tcp_keepalive_time = 7200
net.ipv4.tcp_keepalive_intvl = 75
net.ipv4.tcp_keepalive_probes = 9
net.netfilter.nf_conntrack_tcp_timeout_established = 432000
```

**Comment by titanous** [comment]
We have TCP keep-alives on since `v20160721.2`, so these connections _should_ be cleaned up eventually if the other side goes away. Can you run those commands again so that I can see whether any of the connections have been cleaned up?

**Comment by scottrobertson** [comment]
It does not look like things are getting cleaned up: https://gist.github.com/scottrobertson/383dddc002533e1fb4a314a69f983014

**Comment by johnnagro** [comment]
here is my output to `netstat -tuna | sed -r 's/(([0-9]{1,3}\.){3})[0-9]{1,3}/\1x/g'`

https://gist.github.com/johnnagro/4c071372e1b187cd0e51ac2861584d64

**Comment by titanous** [comment]
This was fixed in #4537
